File: double_circular.py
class Node:
    __slots__='_element',"_prev","_next"
    def __init__(self,element,prev,next):
        self._element=element
        self._prev=prev
        self._next=next
    
class DoubleCircular:
    def __init__(self):
        self.head=None
        self.tail=None
        self.size=0
    
    def __len__(self):
        return self.size
    
    def isEmpty(self):
        return self.size==0
    
    def addLast(self,element):
        newest=Node(element,None,None)
        if self.isEmpty():
            newest._next=self.head
            self.head=newest
            self.head._prev=self.tail
            
        else:
            newest._next=self.tail._next
            self.tail._next=newest
            newest._prev=self.tail
        self.tail=newest
        self.size+=1
    def addFirst(self,element):
        newest=Node(element,None,None)
        if self.isEmpty():
            newest._next=self.head
            self.head=newest
            newest._prev=self.head
        else:
            newest._next=self.head
            self.head=newest
            self.head._prev=self.tail._next
        self.tail=newest
        self.size+=1
    

    def delFirst(self):
        self.tail._next=self.head._next
        self.head=self.head._next
        self.head._prev=self.tail
        self.size-=1

    def delLast(self):
        p=self.head
        i=1
        while i<len(self):
            # print(p._element)
            p=p._next
            i+=1
        self.head._prev=p._next
        p._next=self.head
        self.size-=1
    
    def insertAt(self,element,pos):
        newest=Node(element,None,None)
        if pos>len(self):
            print("please lower the position")
        else:
            p=self.head
            i=1
            while i<(pos-1):
                print(p._element)
                p=p._next
                i+=1
            newest._next=p._next
            p._next=newest
            newest._prev=p
            self.size+=1
    def deleteAny(self,pos):
        p=self.head
        i=1
        while i<(pos-1):
            p=p._next
            i+=1
        p._next=p._next._next
        p._next._prev=p
        self.size-=1

    def display(self):
        p=self.head
        i=0
        while i<len(self):
            print(p._element,end="--->>")
            p=p._next
            i+=1
        print()


dc=DoubleCircular()
dc.addLast(1)
dc.addLast(2)
dc.addLast(3)
dc.addLast(4)
dc.addLast(5)
dc.display()
print("*"*20)
dc.delFirst()
dc.display()
dc.delLast()
dc.display()
dc.insertAt(23,2)
dc.display()
dc.deleteAny(2)
dc.display()
    

File: double_circular.go
package main

import "fmt"

type Node struct {
	prev  *Node
	value int
	next  *Node
}
type DoubleCircular struct {
	head *Node
	tail *Node
	size int
}

func (dc *DoubleCircular) addLast(val int) {
	newNode := &Node{value: val}
	if dc.size == 0 {
		newNode.next = newNode
		dc.head = newNode
		dc.head.prev = newNode
	} else {
		newNode.next = dc.tail.next
		dc.tail.next = newNode
		newNode.prev = dc.tail
	}
	dc.tail = newNode
	dc.size++
}

func (dc *DoubleCircular) display() {
	ptr := dc.head
	for i := 0; i < dc.size; i++ {
		fmt.Printf("%v--->", ptr)
		ptr = ptr.next
	}
	fmt.Println()
}

func (dc *DoubleCircular) addFirst(val int) {
	newNode := &Node{value: val}
	if dc.size == 0 {
		newNode.next = dc.head
		dc.head = newNode
		dc.head.prev = newNode
	} else {
		dc.tail.next = newNode
		newNode.next = dc.head
		newNode.prev = dc.tail
		dc.head = newNode

	}
	dc.size++
}

func (dc *DoubleCircular) addBetween(val, pos int) {
	newest := &Node{value: val}
	ptr := dc.head
	for i := 1; i < pos-1; i++ {
		ptr = ptr.next
	}
	newest.next = ptr.next
	newest.prev = ptr
	ptr.next = newest

	dc.size++
}

func (dc *DoubleCircular) delFirst() {
	if dc.size == 0 {
		fmt.Println("list is empty.")
	} else {
		dc.tail.next = dc.head.next
		dc.head = dc.head.next
		dc.head.prev = dc.tail
	}
	dc.size--
}

func (dc *DoubleCircular) delLast() {
	ptr := dc.head
	if dc.size == 0 {
		fmt.Println("list is empty.")
	} else {
		for i := 0; i < dc.size; i++ {
			ptr = ptr.next
		}
		dc.head.prev = ptr.next
		ptr.next = dc.head
	}
	dc.size--
}

func main() {
	dc := DoubleCircular{}
	dc.addLast(1)
	dc.display()
	dc.addLast(2)
	dc.display()
	dc.addLast(3)
	dc.display()
	dc.addLast(4)
	dc.display()
	// dc.addFirst(0)
	// dc.display()
	// dc.addBetween(33, 3)
	// dc.display()
}
